### PR TITLE
entgql: support generate array for custom JSON type

### PR DESCRIPTION
**Pull request by nzlov**
Use:
```
		field.JSON("pause_times", []PauseTime{}).Annotations(
			entgql.Skip(entgql.SkipMutationCreateInput),
			entgql.Skip(entgql.SkipMutationUpdateInput),
		),
```

Gen:
```
  pauseTimes: [PauseTime!]!
```

Need `PauseTime` type to add declaration to your own graphql file.

### CODE REVIEW

File: entgql/schema.go [hunk 1]
```diff
@@ -727,6 +727,9 @@ func (e *schemaGenerator) mapScalar(gqlType string, f *gen.Field, ant *Annotatio
 			// will generate an enum named "TodoStatus".
 			scalar = gqlType + scalar
 		}
+		if strings.HasPrefix(f.Type.String(), "[]") {
```
Comment (a8m) [inline entgql/schema.go:730]: Use `RType.Kind` instead.

Also, can you please add test cases for this in `internal/todo`? Thanks

Comment (nzlov) [reply entgql/schema.go:730]: Sorry.
RType is nil.

**Comment by giautm** [comment]
why do you not use `entgql.Type('[PauseTime!]')`?

**Comment by nzlov** [comment]
> why do you not use `entgql.Type("[PauseTime!]")`?

why must use?

File: entgql/schema.go (hunk 1 above)
Comment (a8m) [reply entgql/schema.go:730]: It's `nil` only for fields without GoType/JSON configured. Use: `f.Type.RType != nil && f.Type.RType.Kind = reflect.Slice`.

Comment (nzlov) [reply entgql/schema.go]: ok

Comment (giautm) [reply entgql/schema.go]: I'm not sure if there is a case when `Kind` is `reflect.Array`, but can we add it as a backup?

**Comment by giautm** [comment]
Hi @nzlov, can you please fix the test? You need to run `go generate ./...` again in the EntGQL directory. Also, add the `custom` field as `Optional` to avoid breaking the currently test-cases.

File: entgql/schema.go (hunk 1 above)
Comment (nzlov) [reply entgql/schema.go]: ok

**Comment by nzlov** [comment]
> Hi @nzlov, can you please fix the test? You need to run `go generate ./...` again in the EntGQL directory. Also, add the `custom` field as `Optional` to avoid breaking the currently test-cases.

ok

**Comment by giautm** [comment]
I saw you forgot to update the gqlgen.yaml with the custom type. So, gqlgen generate another model for it.

**Comment by nzlov** [comment]
> I saw you forgot to update the gqlgen.yaml with the custom type. So, gqlgen generate another model for it.

`gqlgen.yaml` `autobind` no work?

**Comment by giautm** [comment]
> > I saw you forgot to update the gqlgen.yaml with the custom type. So, gqlgen generate another model for it.
> 
> `gqlgen.yaml` `autobind` no work?

https://github.com/ent/contrib/pull/401/files#diff-e367b846d59dfa604afc6c01ffbba09ca015cac77aed55b188658d2e6734a7d5
The config was missing for other example, like `gotype`, `pulid`, `uuid`.

Please try to fix all issue with PR, run `go test ./...` in local, and revert the change for go.mod, and entproto package

File: entgql/schema.go [hunk 2]
```diff
@@ -727,6 +727,12 @@ func (e *schemaGenerator) mapScalar(gqlType string, f *gen.Field, ant *Annotatio
 			// will generate an enum named "TodoStatus".
 			scalar = gqlType + scalar
 		}
+		if f.Type.RType != nil {
+			switch f.Type.RType.Kind {
+			case reflect.Slice, reflect.Array:
+				scalar = "[" + scalar + "!]"
```
Comment (a8m) [inline entgql/schema.go:733]: The "!" (not null) needs to be inferred from the type. Unlike `T`, `*T` is nullable.

Comment (a8m) [reply entgql/schema.go:733]: @giautm, your thoughts on it? Users can always fallback to the `Type()` option.

Comment (giautm) [reply entgql/schema.go:733]: > The "!" (not null) needs to be inferred from the type. Unlike `T`, `*T` is nullable.

agree.

**Comment by nzlov** [comment]
I can't do the rest of my work because of the coronavirus

**Comment by giautm** [comment]
> I can't do the rest of my work because of the coronavirus

Oh, Hope you getting better soon. I will take care the PR from here.

**Comment by nzlov** [comment]
> > I can't do the rest of my work because of the coronavirus
> 
> Oh, Hope you getting better soon. I will take care the PR from here.

Thx.
I back.

File: entgql/schema.go (hunk 2 above)
Comment (nzlov) [reply entgql/schema.go]: How to fix?

**Comment by nzlov** [comment]
@giautm Did the test fail because the tool version was too high and needed to be downgraded?

**Comment by giautm** [comment]
Just revert the change in entproto

**Comment by giautm** [comment]
Thank @nzlov for your contribution and patience. Wish you always healthy, and to make many contributions to the community in the future. Best regards, ❤️

**Review by giautm — APPROVED** [review]

**Comment by nzlov** [comment]
@giautm thx.

**Comment by kbumsik** [comment]
Wow I just needed this feature and was thinking to contribute it without knowing this PR.
Thank you very much for your work!

**Pull request merged**